File: ruby/problem_5.rb
#!/bin/env ruby
#
# Problem 5
#
# Author::    mattt
#
require 'ruby/euler_lib'

timed_run{

  nums = (1..20).to_a

  loop do
    break if nums.length == 1

    first_num = nums.pop
    next_num = nums.pop

    least_common_multiple = lcm(first_num, next_num)
    nums.push least_common_multiple

  end

  #232792560
  puts nums.shift
}

File: golang/problem_5.go
package main

func problem5() int {
	nums := []int{
		1,
		2,
		3,
		4,
		5,
		6,
		7,
		8,
		9,
		10,
		11,
		12,
		13,
		14,
		15,
		16,
		17,
		18,
		19,
		20,
	}

	var firstNum int
	var nextNum int

	var leastCommonMultiple int

	pop := func() int {
		last := len(nums) - 1

		num := nums[last]
		nums = nums[:last]
		return num
	}

	for {
		if len(nums) == 1 {
			break
		}

		firstNum = pop()
		nextNum = pop()

		leastCommonMultiple = lcm(firstNum, nextNum)
		nums = append(nums, leastCommonMultiple)
	}

	return nums[0]
}
